### PR TITLE
feat: registers sd-transforms; includes refs in source-tokens-only output; resolves refs warnings

### PR DESCRIPTION
**Pull request by adamstankiewicz**
# Description

1. Integrates `style-dictionary` with `sd-transforms` ([GitHub](https://github.com/tokens-studio/sd-transforms)) from [Tokens Studio](https://tokens.studio/). Enables support for copying exported tokens from the Tokens Studio plugin in Figma and adding them to the Paragon tokens schema.
1. Registers a custom transform group `paragon-css` to abstract away our custom transforms, `sd-transforms`, and the default CSS transforms. The previously excluded `size/rem` filter for the default CSS transforms is added back in.
1. Resolves noisy warnings regarding missing output references in resulting token transforms, by now including the reference token(s) in the output. See screenshots below.
    * In Paragon, this use case applies when transforming `light` theme variant tokens that reference `core` tokens. Now, these handful referenced `core` tokens are included in the `light` theme CSS variables, although duplicative of their definitions in `core.css`; this is an acceptable tradeoff to remove noisy, disruptive warnings as it only expands the contents of output files by the number of _referenced_ tokens.
    * In brand packages (e.g., `@edx/elm-theme`), this use case applies when either `core` and/or `light` tokens reference base, default Paragon tokens without redefining the base token within the brand package. Now, the referenced default Paragon token(s) are included in the brand package output.
1. Adds `build-tokens:watch` NPM script to automatically execute the `build-tokens` CLI on file changes, enabling more rapid development when changing tokens and verifying output without having to manually re-run `build-tokens` yourself.

### Resolves output references warning in build output

Related to (3) above. Paragon tokens output on left; brand package tokens output on right (see related `@edx/elm-theme` [PR](https://github.com/edx/elm-theme/pull/24)).

#### Before

image

#### After

image

## Merge Checklist

* [ ] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [ ] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [ ] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [ ] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [ ] Were your changes tested in the `example` app?
* [ ] Is there adequate test coverage for your changes?
* [ ] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please request an a11y review for the PR in the [#wg-paragon](https://openedx.slack.com/archives/C02NR285KV4) Open edX Slack channel.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@openedx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

### CODE REVIEW

File: lib/build-tokens.js [hunk 1]
```diff
@@ -82,7 +84,6 @@ async function buildTokensCommand(commandArgs) {
             },
           },
         ],
-        transforms: StyleDictionary.hooks.transformGroups.css.filter(item => item !== 'size/rem').concat('color/sass-color-functions', 'str-replace'),
```
Comment (adamstankiewicz) [inline lib/build-tokens.js:85]: [inform] abstracted into `transformGroup: 'paragon-css'`

File: styles/css/core/variables.css [hunk 2]
```diff
@@ -36,26 +36,26 @@
   --pgn-transition-base: all .2s ease-in-out; /* Generic transition for any property change */
   --pgn-transition-progress-bar-bar-transition: width .6s ease;
   --pgn-transition-progress-bar-bar-animation-timing: 1s linear infinite;
-  --pgn-transition-form-control: background-color .15s ease-in-out, border-color .15s ease-in-out, box-shadow .15s ease-in-out;
-  --pgn-transition-form-input: border-color .15s ease-in-out, box-shadow .15s ease-in-out;
+  --pgn-transition-form-control: background-color 0.15s ease-in-out, border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
```
Comment (adamstankiewicz) [inline styles/css/core/variables.css:39]: [inform] `sd-transforms` includes a transform to add leading zeros where previously missing.

File: styles/css/core/variables.css [hunk 3]
```diff
@@ -36,26 +36,26 @@
...
+  --pgn-typography-line-height-micro: 0.938rem; /* Micro utils line height. */
   --pgn-typography-line-height-sm: 1.5; /* Small line height. */
   --pgn-typography-line-height-lg: 1.5; /* Large line height. */
   --pgn-typography-line-height-base: 1.5556; /* Basic line height. */
   --pgn-typography-font-weight-display-4: var(--pgn-typography-font-weight-bold); /* Font weight of display level 4. */
   --pgn-typography-font-weight-display-3: var(--pgn-typography-font-weight-bold); /* Font weight of display level 3. */
   --pgn-typography-font-weight-display-2: var(--pgn-typography-font-weight-bold); /* Font weight of display level 2. */
   --pgn-typography-font-weight-display-1: var(--pgn-typography-font-weight-bold); /* Font weight of display level 1. */
-  --pgn-typography-font-weight-table-th: bold; /* Table th font weight. */
+  --pgn-typography-font-weight-table-th: 700; /* Table th font weight. */
```
Comment (adamstankiewicz) [inline styles/css/core/variables.css:58]: [inform] `sd-transforms` changes `bold` to the corresponding, explicit font weight.

File: lib/build-tokens.js [hunk 4]
```diff
@@ -56,9 +56,11 @@ async function buildTokensCommand(commandArgs) {
     source: tokensSource
       ? [`${tokensSource}/core/**/*.json`, `${tokensSource}/core/**/*.toml`]
       : [],
+    preprocessors: ['tokens-studio'],
```
Comment (adamstankiewicz) [inline lib/build-tokens.js:59]: [inform] See `sd-transforms` README for more information on the [`tokens-studio` preprocessor](https://github.com/tokens-studio/sd-transforms?tab=readme-ov-file#using-the-preprocessor).

File: lib/build-tokens.js [hunk 5]
```diff
@@ -56,9 +57,14 @@ async function buildTokensCommand(commandArgs) {
...
+    expand: {
+      typesMap: (await getTokensStudioTransforms()).expandTypesMap,
```
Comment (adamstankiewicz) [inline lib/build-tokens.js:62]: [inform] See `sd-transforms` README for more information on [`expandTypesMap`](https://github.com/tokens-studio/sd-transforms?tab=readme-ov-file#using-expand) for `expand`.

File: tokens/src/core/components/Card.json [hunk 6]
```diff
@@ -15,7 +15,7 @@
       },
       "columns": {
         "margin": { "source": "$card-columns-margin", "$value": "{spacing.card.spacer.y}" },
-        "count": { "source": "$card-columns-count", "$value": "3" },
+        "count": { "source": "$card-columns-count", "$value": "3", "$type": "number" },
```
Comment (adamstankiewicz) [inline tokens/src/core/components/Card.json:18]: Bug fix around `$type` for a _number_ value, as this token does not define a dimension type currently defined above.

File: lib/build-tokens.js [hunk 7]
```diff
@@ -56,9 +57,14 @@ async function buildTokensCommand(commandArgs) {
     source: tokensSource
       ? [`${tokensSource}/core/**/*.json`, `${tokensSource}/core/**/*.toml`]
       : [],
+    preprocessors: ['annotate-tokens-with-references', 'tokens-studio'],
```
Comment (adamstankiewicz) [inline lib/build-tokens.js:60]: [inform] See docs on [preprocessors](https://styledictionary.com/reference/hooks/preprocessors/). `annotate-tokens-with-references` is custom.

See `sd-transforms` README for more information on the [`tokens-studio` preprocessor](https://github.com/tokens-studio/sd-transforms?tab=readme-ov-file#using-the-preprocessor).

File: tokens/style-dictionary.js [hunk 8]
```diff
@@ -330,13 +453,6 @@ const initializeStyleDictionary = async ({ themes }) => {
   ];
   paragonFilters.forEach(({ name, filter, opts }) => registerStyleDictionaryFilter(name, filter, opts));
 
-  themes.forEach((themeVariant) => {
-    StyleDictionary.registerFilter({
-      name: `isThemeVariant.${themeVariant}`,
-      filter: token => token.filePath.includes(themeVariant),
-    });
-  });
-
```
Comment (adamstankiewicz) [inline tokens/style-dictionary.js:339]: [inform] Duplicate. Already defined in `paragonFilters` above.

File: tokens/style-dictionary.js [hunk 9]
```diff
@@ -296,15 +406,28 @@ const initializeStyleDictionary = async ({ themes }) => {
...
    */
   function registerStyleDictionaryFilter(name, filter, filterOptions = {}) {
-    StyleDictionary.registerFilter({ name, filter });
+    StyleDictionary.registerFilter({
+      name,
+      filter: (token, opts) => {
+        const paragonExtensions = token.$extensions?.[paragonExtensionsNamespace];
+        const isReferencedToken = paragonExtensions?.isReferencedToken;
+        const baseFilterResult = typeof filter === 'function' ? filter(token, opts) : filter;
+        return baseFilterResult || isReferencedToken;
```
Comment (adamstankiewicz) [inline tokens/style-dictionary.js:415]: Each custom registered filter will rely on its own `filter` definition _OR_ fallback to whether the token is a reference token (i.e., another token uses it in its `$value` as an alias). Same for the `themeVariant` specific version of the custom filter.

This change makes it possible to include CSS variable definitions in output files that are otherwise not defined, despite previously intentionally being excluded due to being assumed to exist in other output files. Doing so removes the somewhat alarming warning messages that previously were emitted about having output references without their definitions in output files. With this change, these warnings go away in both Paragon and brand packages.

File: lib/build-tokens.js [hunk 10]
```diff
@@ -56,9 +57,14 @@ async function buildTokensCommand(commandArgs) {
     source: tokensSource
       ? [`${tokensSource}/core/**/*.json`, `${tokensSource}/core/**/*.toml`]
       : [],
+    preprocessors: ['pgn-annotate-token-extensions-with-references', 'tokens-studio'],
```
Comment (adamstankiewicz) [inline lib/build-tokens.js:60]: [inform] See docs on [preprocessors](https://styledictionary.com/reference/hooks/preprocessors/). `pgn-annotate-tokens-with-references` is custom.

See `sd-transforms` README for more information on the [`tokens-studio` preprocessor](https://github.com/tokens-studio/sd-transforms?tab=readme-ov-file#using-the-preprocessor).

File: tokens/src/core/global/breakpoints.json [hunk 11]
```diff
@@ -3,7 +3,7 @@
     "$type": "dimension",
     "breakpoint": {
       "xs": {
-        "$value": "0",
+        "$value": "0px",
```
Comment (adamstankiewicz) [inline tokens/src/core/global/breakpoints.json:6]: [inform] Ensures consistency in unit with other breakpoints.

File: lib/build-tokens.js (hunk 5 above)
Comment (adamstankiewicz) [reply lib/build-tokens.js:62]: Related to https://github.com/openedx/paragon/pull/3222

File: tokens/utils.js [hunk 12]
```diff
@@ -2,6 +2,131 @@ const fs = require('fs');
...
+          return;
+        }
+
+        // Mark the referenced token
+        referencedToken.$extensions = {
+          ...referencedToken.$extensions,
+          'org.openedx.paragon': {
+            ...referencedToken.$extensions?.['org.openedx.paragon'],
+            [propertyName]: true,
+          },
```
Comment (adamstankiewicz) [inline tokens/utils.js:81]: See https://tr.designtokens.org/format/#extensions for more details about `$extensions`. It's being used in this implementation to help include references tokens in build output, even if the original filtering doesn't apply.

E.g.,

```js
{
  // token metadata like $value, etc.
  $extensions: {
    'org.openedx.paragon': {
      isReferencedBySourceToken: bool, // useful to include referenced non-source tokens alongside source-only tokens
      isReferencedByThemeVariant: bool, // useful to include referenced non-theme variant tokens alongside theme variant tokens
    },
  },
}
```

**Review by brian-smith-tcril — COMMENTED** [review]
Overall this looks great! I'd be a bit more confident if there were some tests for the new functions in `utils.js`, but I'm on the fence about that being a blocking requirement. There might be other things that have already landed in the design tokens branch that might be missing tests, and in that case I think making an issue to add tests and just adding this new functions to the list of tests to add could work.

tl;dr -
* Looks good!
* Tests?
* Should probably comb through all of the new design tokens functionality and see if there are untested places we should add tests (out of scope for this PR)

**Comment by adamstankiewicz** [comment]
@brian-smith-tcril Thanks for the review! Agreed the lack of tests isn't ideal to ensure confidence in changes. However, as far as I can tell, there are no current tests for any of the logic defined in the `tokens` folder (inclusive of but not limited to the `utils.js` file), e.g. the base `style-dictionary` config, custom transforms/formatters, etc. Given this, I agree with the sentiment to tackle tests as a standalone issue/task separate from this PR.

**Comment by brian-smith-tcril** [comment]
I made an issue for adding tests https://github.com/openedx/paragon/issues/3229, we can add to it as we find places where tests are missing.

**Review by brian-smith-tcril — APPROVED** [review]
:shipit:

**Review by PKulkoRaccoonGang — APPROVED** [review]
